### PR TITLE
Migrate MOTPE sampler

### PR DESCRIPTION
**Pull request by not522**
## Motivation
Part of works for #1980. This PR migrates `MOTPESampler` to the standard samplers module.
## Description of the changes
- add `optuna.samplers.MOTPESampler` copied from `optuna.multi_objective.samplers.MOTPEMultiObjectiveSampler`
- add `optuna.samplers._motpe_hypervolume` copied from `optuna.multi_objective._hypervolume`
- remove "MultiObjective" from sampler's name
- use ordinal `Study` and `FrozenTrial` in `MOTPESampler`
- copy tests

### CODE REVIEW

**Review by HideakiImamura — COMMENTED** [review]
Thanks for the PR! In my opinion, it is better to leave `optuna/multi_objective/_hypervolume` untouched and use it from `optuna/samplers/motpe.py`. It is the same for the test files.

As a follow-up, I think it's better to move and rename `optuna/_multi_objective.py` to `optuna/multi_objective/_utils.py`.

**Comment by not522** [comment]
Thank you. You mean `optuna.multi_objective._hypervolume` will not be deprecated, is it right?

**Comment by HideakiImamura** [comment]
Yes, it is because the `_hypervolume` module is private. I am saying the opposite of what I said before. Sorry for the confusion.

File: optuna/samplers/_motpe.py [hunk 1]
```diff
@@ -0,0 +1,627 @@
...
+        study: optuna.study.Study,
+        trial: optuna.trial.FrozenTrial,
+        param_name: str,
+        param_distribution: BaseDistribution,
+    ) -> Any:
+        if len(study.directions) < 2:
+            raise ValueError(
+                "Number of objectives must be >= 2. "
+                "Please use optuna.samplers.TPESampler for single-objective optimization."
+            ) from None
```
Comment (hvy) [inline optuna/samplers/_motpe.py:172]: Not introduced by this PR so a minor suggestion. I don't think we're in any error handling context so there's no exception to suppress?
```suggestion
            )
```

File: optuna/samplers/_motpe.py [hunk 2]
```diff
@@ -0,0 +1,627 @@
...
+
+def default_gamma(x: int) -> int:
+    return int(np.floor(0.1 * x))
+
+
+def _default_weights_above(x: int) -> np.ndarray:
+    return np.ones(x)
+
+
+class MOTPESampler(TPESampler):
```
Comment (hvy) [inline optuna/samplers/_motpe.py:35]: Would just like to verify but this sampler is stable and not experimental? C.f. https://github.com/optuna/optuna/commit/909324a09d8a45f07b78422284f04cc8c4ca2349

File: optuna/samplers/_motpe.py [hunk 3]
```diff
@@ -0,0 +1,627 @@
...
+
+    This function collects observation pairs from the complete trials of the study.
+    Pruning is currently not supported.
+    The values for trials that don't contain the parameter named ``param_name`` are set to None.
+
+    Objective values are negated if their directions are maximization and all objectives are
+    treated as minimization in the MOTPE algorithm.
+    """
+
+    trials = study._storage.get_all_trials(study._study_id, deepcopy=False)
```
Comment (hvy) [inline optuna/samplers/_motpe.py:606]: Can be a follow up but you can optimize/simplify this and the following code using the following.
```suggestion
    trials = study._storage.get_all_trials(study._study_id, deepcopy=False, states=(optuna.trial.TrialState.COMPLETE,))
```

File: optuna/samplers/_motpe.py [hunk 4]
```diff
@@ -0,0 +1,627 @@
...
+                for i in range(1, num_variables + 1):
+                    x.append(trial.suggest_float(f"x{i}", 0.0, 2.0 * i))
+                return x
+
+
+            sampler = optuna.samplers.MOTPESampler(
+                n_startup_trials=n_startup_trials, n_ehvi_candidates=24, seed=seed
+            )
+            study = optuna.create_study(directions=["minimize"] * num_variables, sampler=sampler)
+            study.optimize(objective, n_trials=250)
```
Comment (hvy) [inline optuna/samplers/_motpe.py:111]: Sounds like a of trials for a doctest example.
```suggestion
            study.optimize(objective, n_trials=n_startup_trials + 10)
```

File: optuna/samplers/_motpe.py [hunk 5]
```diff
@@ -0,0 +1,627 @@
...
+        However, the current implementation only supports initialization with random sampling.
+
+    Example:
+
+        .. testcode::
+
+            import optuna
+
+            seed = 128
+            num_variables = 9
```
Comment (hvy) [inline optuna/samplers/_motpe.py:96]: Related to previous comment about `n_trials`.
```suggestion
            num_variables = 2
```

**Review by hvy — CHANGES_REQUESTED** [review]
Thanks a lot for the PR. Haven't gone through the unit tests yet but left early comments on everything else.

In addition to the comments, I'm wondering how come we don't put this sampler file inside the `_tpe` directory. It seems  related deriving the existing TPE sampler and reusing helpers inside `_tpe`.

File: optuna/samplers/_motpe.py (hunk 2 above)
Comment (hvy) [reply optuna/samplers/_motpe.py:35]: @HideakiImamura , @toshihikoyanase you probably know better than myself given #1530. Do you have any comments on this?

Comment (toshihikoyanase) [reply optuna/samplers/_motpe.py:35]: I'm not pretty sure, but MoTPE may be able to be merged with TPE. Such a change may affect the public interface, and I think it is reasonable to keep the MoTPE experimental.

Comment (HideakiImamura) [reply optuna/samplers/_motpe.py:35]: I agree with @toshihikoyanase. As a follow-up task, we can merge the multi-objective TPE with TPE after the v2.4.0 release.

**Review by HideakiImamura — APPROVED** [review]
Thanks for the update. LGTM.

File: optuna/samplers/_tpe/multi_objective_sampler.py [hunk 6]
```diff
@@ -0,0 +1,628 @@
...
+
+def default_gamma(x: int) -> int:
+    return int(np.floor(0.1 * x))
+
+
+def _default_weights_above(x: int) -> np.ndarray:
+    return np.ones(x)
+
+
+@experimental("2.3.0")
```
Comment (hvy) [inline optuna/samplers/_tpe/multi_objective_sampler.py:36]: This version should be bumped to match our upcoming release.
```suggestion
@experimental("2.4.0")
```

**Review by hvy — APPROVED** [review]
LGTM!

**Pull request merged**